src/Features/LanguageServer/Protocol/ExternalAccess/Razor/SemanticTokensRangesHandler.cs

```diff
@@ -0,0 +1,149 @@
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.ExternalAccess.Razor;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(SemanticTokensRangesHandler)), Shared]
+[Method(SemanticRangesMethodName)]
+internal class SemanticTokensRangesHandler : ILspServiceRequestHandler<SemanticTokensRangesParams, SemanticTokens>
+{
+    private const int TokenSize = 5;
+    public const string SemanticRangesMethodName = "textDocument/semanticTokens/ranges";
+    private readonly IGlobalOptionService _globalOptions;
+    private readonly SemanticTokensRefreshQueue _semanticTokenRefreshQueue;
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public SemanticTokensRangesHandler(
+        IGlobalOptionService globalOptions,
+        SemanticTokensRefreshQueue semanticTokensRefreshQueue)
+    {
+        _globalOptions = globalOptions;
+        _semanticTokenRefreshQueue = semanticTokensRefreshQueue;
+    }
+
+    public async Task<SemanticTokens> HandleRequestAsync(
+            SemanticTokensRangesParams request,
+            RequestContext context,
+            CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(request.TextDocument, "TextDocument is null.");
+        if (request.Ranges.Length == 0)
+        {
+            return new SemanticTokens() { Data = Array.Empty<int>() };
+        }
+
+        var contextDocument = context.GetRequiredDocument();
+        var document = contextDocument.WithFrozenPartialSemantics(cancellationToken);
+        var project = document.Project;
+        var options = _globalOptions.GetClassificationOptions(project.Language) with { ForceFrozenPartialSemanticsForCrossProcessOperations = true };
+        var capabilities = context.GetRequiredClientCapabilities();
+        using var _ = ArrayBuilder<int[]>.GetInstance(request.Ranges.Length, out var responseData);
+        foreach (var range in request.Ranges)
+        {
+            var semanticTokens = await SemanticTokensHelpers.ComputeSemanticTokensDataAsync(
+                capabilities,
+                document,
+                range,
+                options,
+                cancellationToken).ConfigureAwait(false);
+
+            if (semanticTokens is not null && semanticTokens.Length != 0)
+            {
+                responseData.Add(semanticTokens);
+            }
+        }
+
+        await _semanticTokenRefreshQueue.TryEnqueueRefreshComputationAsync(project, cancellationToken).ConfigureAwait(false);
```

Comment:
Wasn't sure if this refresh queue is necessary. Replicated usage from `SemanticTokensRangeHandler`:

https://github.com/dotnet/roslyn/blob/91f67f373d2e0b65e4a43203f2962b200804e63e/src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs#L70

Reply:
Yeah, we should match the behaviour, it's good.